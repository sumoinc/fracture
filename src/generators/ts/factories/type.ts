import { SyntaxKind, addSyntheticLeadingComment, factory } from "typescript";
import { TypeAttributeFactory } from "./type-attribute";
import { TypescriptFactory } from "./typescript-factory";
import { FractureService } from "../../../core";
import { Structure } from "../../../core/structure";
import { TypescriptStrategy } from "../strategy";

export class TypeFactory extends TypescriptFactory {
  /**
   * Turns and array of Structures into Typescript types.
   */
  public static toTypes({
    service,
    structures,
  }: {
    service: FractureService;
    structures: Structure[];
  }) {
    return structures.map((structure) => {
      return TypeFactory.toType({
        service,
        structure,
      });
    });
  }

  /**
   * Turns one Structure into a Typescript type.
   */
  public static toType({
    service,
    structure,
  }: {
    service: FractureService;
    structure: Structure;
  }) {
    /***************************************************************************
     * Get Strategy
     **************************************************************************/

    const strategy = TypescriptStrategy.of(service);

    if (!strategy) {
      throw new Error(`No TypescriptStrategy defined at the service level.`);
    }

    /***************************************************************************
     * Build Type
     **************************************************************************/

    // in case of MyType<T> or similar
    const typeParameter = structure.typeParameter
      ? [
          factory.createTypeParameterDeclaration(
            undefined,
            factory.createIdentifier(structure.typeParameter),
            undefined,
            undefined
          ),
        ]
      : undefined;

    // build properties for this type
    const properties = TypeAttributeFactory.toTypeProperies({
      service,
      attributes: structure.attributes,
    });

    // define the type
    const type = factory.createTypeAliasDeclaration(
      [factory.createToken(SyntaxKind.ExportKeyword)],
      factory.createIdentifier(strategy.formatTypeName(structure.name)),
      typeParameter,
      properties
    );

    // add comments if needed
    if (structure.comments.length) {
      addSyntheticLeadingComment(
        type,
        SyntaxKind.MultiLineCommentTrivia,
        `*\n * ${structure.comments.join("\n * ")}\n`,
        true
      );
    }

    return type;
  }
}
