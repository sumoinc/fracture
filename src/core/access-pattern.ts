import { deepMerge } from "projen/lib/util";
import { ValueOf } from "type-fest";
import { Operation, OPERATION_SUB_TYPE } from "./operation";
import { Resource } from "./resource";
import {
  ResourceAttribute,
  ResourceAttributeGenerator,
  ResourceAttributeOptions,
} from "./resource-attribute";
import { FractureComponent } from "../core/component";
import { DynamoGsi } from "../dynamodb/dynamo-gsi";
import { DynamoTable } from "../dynamodb/dynamo-table";

export const ACCESS_PATTERN_TYPE = {
  /**
   * Main PK and PK pattern used to identify this record.
   */
  IDENTIFIER: "Identifier",
  /**
   * Lookup pattern for quick begins_with() searching and such.
   */
  LOOKUP: "lookup",
  /**
   * Any other AP
   */
  OTHER: "other",
} as const;

export interface AccessPatternOptions {
  dynamoGsi?: DynamoGsi;
  pkAttributeOptions?: ResourceAttributeOptions;
  skAttributeOptions?: ResourceAttributeOptions;
  type?: ValueOf<typeof ACCESS_PATTERN_TYPE>;
}

export class AccessPattern extends FractureComponent {
  // member components
  public pkAttribute: ResourceAttribute;
  public skAttribute: ResourceAttribute;
  public operations: Operation[] = [];
  // parent
  public readonly resource: Resource;
  // all other options
  public readonly options: Required<AccessPatternOptions>;

  constructor(resource: Resource, options: AccessPatternOptions) {
    super(resource.fracture);

    /***************************************************************************
     *
     * DEFAULT OPTIONS
     *
     **************************************************************************/

    // make a new gsi is one wasn't passed into the Access Pattern
    const dynamoGsi = options.dynamoGsi
      ? options.dynamoGsi
      : new DynamoGsi(resource.dynamoTable);

    const defaultOptions: Required<AccessPatternOptions> = {
      dynamoGsi,
      pkAttributeOptions: {
        name: dynamoGsi.pkName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      skAttributeOptions: {
        name: dynamoGsi.skName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      type: ACCESS_PATTERN_TYPE.OTHER,
    };

    /***************************************************************************
     *
     * INIT ACCESS PATTERN
     *
     **************************************************************************/

    // all other options
    this.options = deepMerge([
      defaultOptions,
      options,
    ]) as Required<AccessPatternOptions>;
    // the attribute might already be defined.
    const pkAttribute = resource.getAttributeByName(dynamoGsi.pkName);
    const skAttribute = resource.getAttributeByName(dynamoGsi.skName);

    // member components
    this.pkAttribute = pkAttribute
      ? pkAttribute
      : new ResourceAttribute(resource, this.options.pkAttributeOptions);
    this.skAttribute = skAttribute
      ? skAttribute
      : new ResourceAttribute(resource, this.options.skAttributeOptions);

    // parent + inverse
    this.resource = resource;
    this.resource.accessPatterns.push(this);

    this.project.logger.info(
      `Access Pattern for: "${this.pkAttribute.name}:${this.skAttribute.name}" initialized.`
    );

    return this;
  }

  public build() {
    this.project.logger.debug(
      `BUILD Access Pattern: "${this.pkAttribute.name}:${this.skAttribute.name}" called.`
    );
    /***************************************************************************
     *
     * CRUD OPERATIONS
     *
     * TODO
     *
     **************************************************************************/
    /*
    new Operation(this, {
      operationType: OPERATION_TYPE.MUTATION,
      operationSubType: OPERATION_SUB_TYPE.LIST,
    });
    */
  }

  addPkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.pkAttribute.compositionSources.push(sourceAttribute);
  }

  addSkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.skAttribute.compositionSources.push(sourceAttribute);
  }

  /**
   * Returns operation for given sub-type
   */
  public getOperation(
    operationSubType: ValueOf<typeof OPERATION_SUB_TYPE>
  ): Operation {
    const returnOperation = this.operations.find(
      (o) => o.operationSubType === operationSubType
    );

    if (!returnOperation) {
      throw new Error(`Operation not found for sub-type: ${operationSubType}`);
    }

    return returnOperation;
  }

  public get createOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.CREATE_ONE);
  }

  public get readOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.READ_ONE);
  }

  public get updateOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.UPDATE_ONE);
  }

  public get deleteOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.DELETE_ONE);
  }

  public get isKeyAccessPattern(): boolean {
    return this.dynamoTable.keyDynamoGsi === this.dynamoGsi;
  }

  public get isLookupAccessPattern(): boolean {
    return this.dynamoTable.lookupDynamoGsi === this.dynamoGsi;
  }

  public get dynamoTable(): DynamoTable {
    return this.dynamoGsi.dynamoTable;
  }

  public get dynamoGsi(): DynamoGsi {
    return this.options.dynamoGsi;
  }
}
