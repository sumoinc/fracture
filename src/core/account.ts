import { Component } from "projen";
import { Organization } from "./organization";
import { OrganizationalUnit } from "./organizational-unit";
import { Region, RegionOptions } from "./region";

export interface AccountOptions {
  id: string;
  name?: string;
  rootEmail?: string;
  isManagementAccount?: boolean;
  organizationalUnit?: OrganizationalUnit;
}

export class Account extends Component {
  // member components
  public readonly regions: Region[];
  // parent
  public readonly organization: Organization;
  // all other options
  public readonly options: AccountOptions;

  constructor(organization: Organization, options: AccountOptions) {
    super(organization.project);

    /***************************************************************************
     *
     * INIT OPERATION
     *
     **************************************************************************/

    // member components
    this.regions = [];

    // parents + inverse
    this.organization = organization;
    this.organization.accounts.push(this);

    // all other options
    const defaultOptions: Partial<AccountOptions> = {
      isManagementAccount: false,
    };
    this.options = { ...defaultOptions, ...options };

    // if OU defined, add account to OU
    if (this.options.organizationalUnit) {
      this.options.organizationalUnit.accounts.push(this);
    }

    // debugging output
    this.project.logger.info(`INIT Account: "${this.id}"`);
  }

  public get id() {
    return this.options.id;
  }

  public get name() {
    return this.options.name;
  }

  public get rootEmail() {
    return this.options.rootEmail;
  }

  public get isManagementAccount() {
    return this.options.isManagementAccount;
  }

  public get organizationalUnit() {
    return this.options.organizationalUnit;
  }

  /*****************************************************************************
   *
   *  Configuration Helpers
   *
   ****************************************************************************/

  /**
   * Add an region to an account.
   *
   * @param {RegionOptions}
   * @returns {Region}
   */
  public addRegion(options: RegionOptions) {
    const region = this.regions.find((r) => r.id === options.id);
    return region ? region : new Region(this, options);
  }
}
