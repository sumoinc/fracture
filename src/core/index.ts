export * from "./fracture-app";
export * from "./fracture-service";
export * from "./fracture";
export * from "./region";
export * from "./resource";
