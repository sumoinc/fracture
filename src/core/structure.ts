import { deepMerge } from "projen/lib/util";
import { SetRequired, ValueOf } from "type-fest";
import { FractureComponent } from "./component";
import { Operation, OPERATION_SUB_TYPE } from "./operation";
import { Resource } from "./resource";
import { ResourceAttribute } from "./resource-attribute";
import { Service } from "./service";

/******************************************************************************
 * TYPES
 *****************************************************************************/

export type StructureOptions = {
  /**
   *  Type of structure
   */
  type?: ValueOf<typeof STRUCTURE_TYPE>;
  /**
   * Required if type is input or output
   */
  operation?: Operation;
  /**
   * Comment lines to add to the Structure.
   * @default []
   */
  comments?: string[];
};

export const STRUCTURE_TYPE = {
  INPUT: "Input",
  OUTPUT: "Output",
  /**
   * Represents the structure of persistant data
   */
  DATA: "Data",
  /**
   * Represents the structure of transient data such as messages or events
   */
  TRANSIENT: "Transient",
} as const;

/******************************************************************************
 * CLASS
 *****************************************************************************/

export class Structure extends FractureComponent {
  // member components
  // parent
  public readonly resource: Resource;
  // all other options
  public readonly options: SetRequired<StructureOptions, "type" | "comments">;

  constructor(resource: Resource, options: StructureOptions) {
    super(resource.fracture);

    /***************************************************************************
     *
     * DEFAULT OPTIONS
     *
     **************************************************************************/

    const defaultOptions: Partial<StructureOptions> = {
      type: STRUCTURE_TYPE.DATA,
      comments: ["A gereric type"],
    };

    /***************************************************************************
     *
     * INIT OPERATION
     *
     **************************************************************************/

    // member components

    // parents + inverse
    this.resource = resource;
    this.resource.structures.push(this);

    // all other options
    this.options = deepMerge([defaultOptions, options]) as SetRequired<
      StructureOptions,
      "type" | "comments"
    >;

    this.project.logger.info(`INIT Structure: "${this.name}"`);

    //
    if (
      (this.options.type === STRUCTURE_TYPE.INPUT ||
        this.options.type === STRUCTURE_TYPE.OUTPUT) &&
      !this.options.operation
    ) {
      throw new Error(
        `Operation option is required for Input and Output Structires`
      );
    }
  }

  // not used here (yet)
  public build() {
    this.project.logger.info(`BUILD Structure: "${this.name}"`);
  }

  /**
   * Structure name, based on the naming strategy
   */
  public get name() {
    const resourceName = this.options.operation
      ? this.options.operation.name
      : this.resource.name;
    const prefix =
      this.fracture.options.namingStrategy.structures.prefixes[
        this.options.type
      ];
    const suffix =
      this.fracture.options.namingStrategy.structures.suffixes[
        this.options.type
      ];

    return [prefix, resourceName, suffix]
      .filter((part) => part.length > 0)
      .join("-");
  }

  public get type() {
    return this.options.type;
  }

  public getResourceAttributeByName(name: string) {
    return this.attributes.find((a) => a.name === name);
  }

  /**
   * Includes all public attributes, plus generated attributes and GSI
   * attributes. This is a good list to use for dynamodb operations.
   */
  public get attributes() {
    return this.resource.attributes.sort((a, b) => {
      if (a.sortPosition < b.sortPosition) {
        return -1;
      } else if (a.sortPosition > b.sortPosition) {
        return 1;
      } else {
        return 0;
      }
    });
  }

  /**
   * Attributes that should be exposed to the public. Inputs, outputs, etc
   */
  public get publicAttributes(): ResourceAttribute[] {
    switch (this.type) {
      case STRUCTURE_TYPE.DATA:
        return this.attributes.filter((a: ResourceAttribute) => {
          return !a.isAccessPatternKey;
        });

      // INPUTS
      case STRUCTURE_TYPE.INPUT:
        switch (this.operation!.operationSubType) {
          case OPERATION_SUB_TYPE.CREATE_ONE:
            return this.attributes.filter((a: ResourceAttribute) => {
              return a.isData;
            });

          case OPERATION_SUB_TYPE.READ_ONE:
          case OPERATION_SUB_TYPE.DELETE_ONE:
            return this.keyAttributeSources.filter((a) => {
              return !a.isGeneratedOn(this.operation);
            });

          case OPERATION_SUB_TYPE.UPDATE_ONE:
            return this.attributes.filter((a: ResourceAttribute) => {
              return (
                !a.isSystem && (a.isData || !a.isGeneratedOn(this.operation))
              );
            });

          default:
            return [];
        }

      case STRUCTURE_TYPE.OUTPUT:
        return this.attributes.filter((a: ResourceAttribute) => {
          return !a.isAccessPatternKey && a.isOutputOn(this.operation);
        });

      case STRUCTURE_TYPE.TRANSIENT:
        return this.attributes.filter((a: ResourceAttribute) => {
          return !a.isAccessPatternKey;
        });
    }
  }

  public get itemAttributes(): ResourceAttribute[] {
    // only applies to inputs
    if (this.type !== STRUCTURE_TYPE.INPUT) {
      return [];
    }
    // on create, put in everything
    if (this.operation?.operationSubType === OPERATION_SUB_TYPE.CREATE_ONE) {
      return this.attributes;
    }

    // on read we need all the parts of the pk and sk that are not generated on read
    if (this.operation?.operationSubType === OPERATION_SUB_TYPE.READ_ONE) {
      return this.keyAttributeSources.filter((a) => !a.isGenerated);
    }

    return [];
  }

  public get keyAttributes(): ResourceAttribute[] {
    // only applies to inputs
    if (this.type === STRUCTURE_TYPE.INPUT) {
      return this.attributes.filter((a) => a.isPartitionKey || a.isSortKey);
    }
    return [];
  }

  public get keyAttributeSources(): ResourceAttribute[] {
    let returnAttributes = [] as ResourceAttribute[];
    this.keyAttributes.forEach((keyAttribute) => {
      keyAttribute.compositionSources.forEach((sourceAttribute) => {
        returnAttributes.push(sourceAttribute as ResourceAttribute);
      });
    });
    return returnAttributes;
  }

  public get generatedAttributes(): ResourceAttribute[] {
    // only applies to inputs
    if (this.type === STRUCTURE_TYPE.INPUT) {
      return this.attributes.filter((a) => a.isGeneratedOn(this.operation));
    }
    return [];
  }

  public get attributeNames() {
    return this.attributes.map((attribute) => attribute.name);
  }

  public get operation(): Operation | undefined {
    return this.options.operation;
  }

  public get service(): Service {
    return this.resource.service;
  }
}
