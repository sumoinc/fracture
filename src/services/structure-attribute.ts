import { paramCase } from "change-case";
import { Component } from "projen";
import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "./resource-attribute";
import { Service } from "./service";

export const StructureAttributeType = {
  ...ResourceAttributeType,
  ARRAY: "Array",
  CUSTOM: "Custom",
} as const;

export type StructureAttributeOptions = {
  /**
   * Full long name for this attribute.
   *
   * @example 'phone-number'
   */
  name: string;
  /**
   * Type for this structure attribute.
   *
   * @default StructureAttributeType.STRING
   */
  type?: ValueOf<typeof StructureAttributeType>;
  /**
   * Type parameter for this structure attribute.
   *
   * @default any
   * @example 'T' for MyType<T> generic
   */
  typeParameter?: string;
  /**
   * Comment lines to add to the Resource.
   *
   * @default []
   */
  comments?: string[];
  /**
   * Is this attribute required?
   *
   * @default true
   */
  required?: boolean;
};

export class StructureAttribute extends Component {
  /**
   * Full long name for this attribute.
   *
   * @example 'phone-number'
   */
  public readonly name: string;
  /**
   * Type for this structure attribute.
   *
   * @default StructureAttributeType.STRING
   */
  type: ValueOf<typeof StructureAttributeType>;
  /**
   * Type parameter for this structure attribute.
   *
   * @default any
   * @example 'T' for MyType<T> generic
   */
  public readonly typeParameter?: string;
  /**
   * Comment lines to add to the Resource.
   *
   * @default []
   */
  public readonly comments: string[];

  /**
   * Is this attribute required?
   *
   * @default true
   */
  public readonly required: boolean;

  constructor(
    public readonly project: Service,
    options: StructureAttributeOptions
  ) {
    super(project);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.name = paramCase(options.name);
    this.type = options.type ?? StructureAttributeType.STRING;
    this.typeParameter = options.typeParameter ?? "any";
    this.comments = options.comments ?? [];
    this.required = options.required ?? true;

    return this;
  }
}
