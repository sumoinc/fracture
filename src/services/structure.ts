import { paramCase } from "change-case";
import { Component } from "projen";
import { TypeScriptProject } from "projen/lib/typescript/typescript.js";
import { DataService } from "./data-service.js";
import {
  StructureAttribute,
  StructureAttributeOptions,
} from "./structure-attribute";

/******************************************************************************
 * TYPES
 *****************************************************************************/

export type StructureOptions = {
  /**
   *  Name for the structure.
   */
  name: string;

  /**
   * Type parameter for this structure type.
   *
   * @default undefined
   * @example 'T' for MyType<T> generic
   */
  typeParameter?: string;

  /**
   * Comment lines to add to the Structure.
   *
   * @default []
   */
  comments?: string[];

  /**
   * Options for attributes to add when initializing the structure.
   */
  attributeOptions?: Omit<StructureAttributeOptions, "structure">[];
};

/******************************************************************************
 * CLASS
 *****************************************************************************/

export class Structure extends Component {
  /**
   * Returns a structure by name, or undefined if it doesn't exist
   */
  public static byName(
    service: DataService,
    name: string
  ): Structure | undefined {
    const isDefined = (c: Component): c is Structure =>
      c instanceof Structure && c.name === name;
    return service.structures.find(isDefined);
  }

  /**
   * Returns all structures for service
   */
  public static all(project: TypeScriptProject): Structure[] {
    const isDefined = (c: Component): c is Structure => c instanceof Structure;
    return project.components.filter(isDefined);
  }

  /**
   *  Name for the structure.
   */
  public readonly name: string;
  /**
   * Type parameter for this structure type.
   *
   * @default undefined
   * @example 'T' for MyType<T> generic
   */
  public readonly typeParameter?: string;
  /**
   * Comment lines to add to the Structure.
   *
   * @default []
   */
  public readonly comments: string[];
  /**
   * All attributes in this structure.
   */
  public readonly attributes: StructureAttribute[] = [];

  constructor(public readonly project: DataService, options: StructureOptions) {
    super(project);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.name = paramCase(options.name);
    this.typeParameter = options.typeParameter;
    this.comments = options.comments ?? [];

    if (options.attributeOptions) {
      options.attributeOptions.forEach((attributeOption) => {
        this.addAttribute(attributeOption);
      });
    }

    /***************************************************************************
     * Add to Service
     **************************************************************************/

    if (Structure.byName(this.service, this.name)) {
      throw new Error(
        `Resource "${this.service.name}" already has a structure named "${this.name}"`
      );
    }

    this.service.structures.push(this);

    return this;
  }

  /**
   * Add a single structure attribute to the structure.
   */
  public addAttribute(options: Omit<StructureAttributeOptions, "structure">) {
    const attribute = new StructureAttribute(this.project, {
      structure: this,
      ...options,
    });
    this.attributes.push(attribute);
    return attribute;
  }

  public get service() {
    return this.project;
  }
}
