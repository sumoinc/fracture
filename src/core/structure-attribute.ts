import { deepMerge } from "projen/lib/util";
import { ValueOf } from "type-fest";
import { FractureComponent } from "./component";
import { Operation, OPERATION_SUB_TYPE } from "./operation";
import { Resource } from "./resource";
import {
  ResourceAttribute,
  ResourceAttributeGenerator,
  ResourceAttributeOptions,
} from "./resource-attribute";
import { Service } from "./service";
import { Structure } from "./structure";

/******************************************************************************
 * TYPES
 *****************************************************************************/

export const STRUCTURE_ATTRIBUTE_TYPE = {
  PUBLIC: "public",
  PRIVATE: "private",
} as const;

export type StructureAttributeOptions = {
  /**
   * What resource attribute is this structure attribute based on?
   */
  resourceAttribute: ResourceAttribute;
  /**
   * Replacement / override options for structure attribute
   */
  structureAttributeOptions?: Partial<ResourceAttributeOptions>;
  /**
   * Is this a public or private attribute?
   */
  structureAttributeType: ValueOf<typeof STRUCTURE_ATTRIBUTE_TYPE>;
};

/******************************************************************************
 * CLASS
 *****************************************************************************/

export class StructureAttribute extends FractureComponent {
  // member components
  public readonly compositionSources: StructureAttribute[];
  public readonly structureAttributeType: ValueOf<
    typeof STRUCTURE_ATTRIBUTE_TYPE
  >;
  // parent
  public readonly structure: Structure;

  // all other options
  public readonly options: Required<ResourceAttributeOptions>;

  constructor(structure: Structure, options: StructureAttributeOptions) {
    super(structure.fracture);

    /***************************************************************************
     *
     * INIT OPERATION
     *
     **************************************************************************/

    // member components
    this.compositionSources = options.resourceAttribute.compositionSources.map(
      (source) => {
        return new StructureAttribute(structure, {
          resourceAttribute: source,
          structureAttributeType: options.structureAttributeType,
        });
      }
    );
    this.structureAttributeType = options.structureAttributeType;

    // parents + inverse
    this.structure = structure;

    // all other options
    this.options = deepMerge([
      JSON.parse(JSON.stringify(options.resourceAttribute.options)),
      options.structureAttributeOptions ?? {},
    ]) as Required<ResourceAttributeOptions>;
  }

  public get name(): string {
    return this.options.name;
  }

  public get shortName(): string {
    return this.options.shortName;
  }

  /**
   * Boost composed attributes to the end of the list. Doing this
   * ensures helpsmake sure that we have all the other attributes we need prior
   * to building them.
   *
   */
  public get sortPosition(): number {
    const compositionBoost = this.isComposed ? 1000 : 0;
    return this.options.sortPosition + compositionBoost;
  }

  public get isComposed(): boolean {
    return this.hasGenerator(ResourceAttributeGenerator.COMPOSITION);
  }

  public get isRequired() {
    // if the required optioun is set, require it here too
    if (this.options.isRequired) {
      return true;
    }

    // otherwise, false
    return false;
  }

  /**
   *
   * @param generator Does this attribute have a generator for this specific type?
   * @returns
   */
  public hasGenerator(
    generator: ValueOf<typeof ResourceAttributeGenerator>
  ): boolean {
    return (
      this.isGenerated &&
      (this.hasCreateGenerator(generator) ||
        this.hasReadGenerator(generator) ||
        this.hasUpdateGenerator(generator) ||
        this.hasDeleteGenerator(generator) ||
        this.hasImportGenerator(generator))
    );
  }

  // generated fields
  public get isGenerated(): boolean {
    return (
      this.isCreateGenerated ||
      this.isReadGenerated ||
      this.isUpdateGenerated ||
      this.isDeleteGenerated ||
      this.isImportGenerated
    );
  }
  public get isCreateGenerated(): boolean {
    return this.options.createGenerator !== ResourceAttributeGenerator.NONE;
  }
  public get isReadGenerated(): boolean {
    return this.options.readGenerator !== ResourceAttributeGenerator.NONE;
  }
  public get isUpdateGenerated(): boolean {
    return this.options.updateGenerator !== ResourceAttributeGenerator.NONE;
  }
  public get isDeleteGenerated(): boolean {
    return this.options.deleteGenerator !== ResourceAttributeGenerator.NONE;
  }
  public get isImportGenerated(): boolean {
    return this.options.importGenerator !== ResourceAttributeGenerator.NONE;
  }

  public hasCreateGenerator(
    generator: ValueOf<typeof ResourceAttributeGenerator>
  ): boolean {
    return this.isCreateGenerated && this.options.createGenerator === generator;
  }
  public hasReadGenerator(
    generator: ValueOf<typeof ResourceAttributeGenerator>
  ): boolean {
    return this.isReadGenerated && this.options.readGenerator === generator;
  }
  public hasUpdateGenerator(
    generator: ValueOf<typeof ResourceAttributeGenerator>
  ): boolean {
    return this.isUpdateGenerated && this.options.updateGenerator === generator;
  }
  public hasDeleteGenerator(
    generator: ValueOf<typeof ResourceAttributeGenerator>
  ): boolean {
    return this.isDeleteGenerated && this.options.deleteGenerator === generator;
  }
  public hasImportGenerator(
    generator: ValueOf<typeof ResourceAttributeGenerator>
  ): boolean {
    return this.isImportGenerated && this.options.importGenerator === generator;
  }

  public generatorForOperation(operation: Operation) {
    switch (operation.options.operationSubType) {
      case OPERATION_SUB_TYPE.CREATE_ONE:
        return this.options.createGenerator;
      case OPERATION_SUB_TYPE.READ_ONE:
        return this.options.readGenerator;
      case OPERATION_SUB_TYPE.UPDATE_ONE:
        return this.options.updateGenerator;
      case OPERATION_SUB_TYPE.DELETE_ONE:
        return this.options.deleteGenerator;
      case OPERATION_SUB_TYPE.IMPORT_ONE:
        return this.options.importGenerator;
      default:
        throw new Error(`Unknown sub-operation: ${operation}`);
    }
  }

  public get service(): Service {
    return this.resource.service;
  }

  public get resource(): Resource {
    return this.structure.resource;
  }
}
