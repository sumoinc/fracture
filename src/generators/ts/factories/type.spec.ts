import { TypeFactory } from "./type";
import { Fracture, FractureService } from "../../../core";
import { Structure } from "../../../core/structure";
import { StructureAttributeType } from "../../../core/structure-attribute";

let fracture: Fracture;
let service: FractureService;

beforeEach(() => {
  fracture = new Fracture();
  service = new FractureService(fracture, { name: "foo" });
});

test("Smoke test", () => {
  const structure = new Structure(service, {
    name: "MyType",
    attributes: [
      {
        name: "id",
        type: StructureAttributeType.STRING,
      },
    ],
  });
  const type = TypeFactory.toType({ service, structure });
  const content = TypeFactory.print([type]);

  console.log(content);
});
