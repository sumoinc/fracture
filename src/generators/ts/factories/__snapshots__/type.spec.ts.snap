// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All attribute types should match snapshot 1`] = `
"export type MyType = {
    guid: string;
    string: string;
    int: number;
    float: number;
    boolean: boolean;
    date: string;
    time: string;
    dateTime: string;
    timestamp: number;
    email: string;
    json: string;
    phone: string;
    url: string;
    ipaddress: string;
    count: number;
    average: number;
    sum: number;
    array: Array<any>;
    custom: any;
};
"
`;

exports[`Generic and type param support 1`] = `
"export type MyType<T> = {
    arrayType: Array<T>;
    customType: CustomValue;
};
"
`;

exports[`Optional and Required work 1`] = `
"export type MyType = {
    shouldBeRequired: string;
    shouldBeOptional?: string;
};
"
`;

exports[`Smoke test 1`] = `
"export type MyType = {};
"
`;
